### PR TITLE
resolve toc node moniker inside LoadToc function since the document info is lost after LoadToc

### PR DESCRIPTION
**Pull request by 928PJY**
https://ceapex.visualstudio.com/Engineering/_workitems/edit/96568/
Since we resolve the moniker of toc node in a post processor, we just have the href resolved result and lost the document info, we cannot get the accurate moniker resolve result.

Sample:
# toc.md
```
# [A-v1](v1/a.md)
# [A-v2](v2/a.md)
```
In the post processor, we cannot distinguish those two nodes, the moniker resolved result will be the same for those two nodes which have a different version.

there are two solution:
- add document property to TOCItem Model
- resolve moniker inside the LoadToc function

### CODE REVIEW

**Comment by OsmondJiang** [comment]
we already have the mapping between href and document in BuildTableOfContent

File: src/docfx/build/toc/BuildTableOfContents.cs [hunk 1]
```diff
@@ -161,6 +113,14 @@ void ResolveItemMonikers(List<TableOfContentsItem> items)
                     }
 
                     return (link, display, buildItem);
+                },
+                (document) =>
+                {
+                    if (monikerMap != null && document != null && monikerMap.TryGetValue(document, out var monikers))
```
Comment (yufeih) [inline src/docfx/build/toc/BuildTableOfContents.cs]: feels like `MonikerMap` and `MonikerProvider` is the same thing, can we call `MonikerProvider.GetFileLevelMonikers` here instead? Can handle file --> moniker cache internally.

Comment (928PJY) [reply src/docfx/build/toc/BuildTableOfContents.cs]: we cannot get the file metadata here

---
In reply to: [291923815](https://github.com/dotnet/docfx/pull/4702#discussion_r291923815) [](ancestors = 291923815)

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 2]
```diff
@@ -116,30 +136,65 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
                 var topicUid = tocModelItem.Uid;
 
+                var monikers = new List<string>();
+
                 var (resolvedTocHref, resolvedTopicItemFromTocHref, subChildren) = ProcessTocHref(tocHref);
                 var (resolvedTopicHref, resolvedTopicName, document) = ProcessTopicItem(topicUid, topicHref);
 
                 // set resolved href back
-                tocModelItem.Href = resolvedTocHref ?? resolvedTopicHref ?? resolvedTopicItemFromTocHref?.Href;
+                if (resolvedTocHref != null)
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: keep this line un-changed and you can merge resolvedTocHref and resolveTopicHref monikers together?

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 3]
```diff
@@ -116,30 +136,65 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
+                }
+                else if (resolvedTopicHref != null)
+                {
+                    // TopicHref
+                    tocModelItem.Href = resolvedTopicHref;
+
+                    var linkType = UrlUtility.GetLinkType(resolvedTopicHref);
+                    if (linkType == LinkType.External || linkType == LinkType.AbsolutePath)
+                    {
+                        monikers = resolveMoniker(rootPath);
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: Or you can directly return toc's moniker here？ no more extendResolveMoniker?

File: src/docfx/build/toc/BuildTableOfContents.cs (hunk 1 above)
Comment (yufeih) [reply src/docfx/build/toc/BuildTableOfContents.cs]: should be able to get file metadata using `MetadataProvider`

Comment (928PJY) [reply src/docfx/build/toc/BuildTableOfContents.cs]: that method is just for document without yaml header, for now is  redirection and resource file.

---
In reply to: [291939668](https://github.com/dotnet/docfx/pull/4702#discussion_r291939668) [](ancestors = 291939668)

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 4]
```diff
@@ -98,10 +102,19 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
 
             var (errors, model) = LoadTocModel(context, file, content);
 
+            if (file == rootPath)
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: Can this section be lifted outside this recursive function? It makes it easier to see what are applied to every node vs what else we do

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 5]
```diff
@@ -116,30 +129,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
+                    tocModelItem.Href = resolvedTopicItemFromTocHref?.Href;
+                    monikers = resolvedTopicItemFromTocHref?.Monikers;
+                }
+
                 tocModelItem.TocHref = resolvedTocHref;
                 tocModelItem.Name = tocModelItem.Name ?? resolvedTopicName;
                 tocModelItem.Items = subChildren?.Items ?? tocModelItem.Items;
 
+                // Union with children's monikers
+                var childrenMonikers = tocModelItem.Items?.SelectMany(child => child?.Monikers ?? new List<string>()) ?? new List<string>();
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: did we initialize items and monkers to empty list? these ? and ?? null checks are clumsy

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 4 above)
Comment (928PJY) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: filemoniker can only be got after the metadata has been extract

---
In reply to: [292299489](https://github.com/dotnet/docfx/pull/4702#discussion_r292299489) [](ancestors = 292299489)

Comment (yufeih) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: ok, then can this be done after #4711 ?

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 6]
```diff
@@ -21,10 +21,19 @@ internal static class TableOfContentsParser
...
 
+        public delegate List<string> ResolveMoniker(Document doc);
+
         public static (List<Error> errors, TableOfContentsModel model)
-            Load(Context context, Document file, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref)
+            Load(Context context, Document file, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref, ResolveMoniker resolveMoniker)
         {
-            return LoadInternal(context, file, file, resolveContent, resolveHref, resolveXref, new List<Document>());
+            var errors = new List<Error>();
+            var (monikerError, fileMonikers) = context.MonikerProvider.GetFileLevelMonikers(file);
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: will directly use `resolveMoniker` here?

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 7]
```diff
@@ -101,7 +112,7 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
             if (model.Items.Count > 0)
             {
                 parents.Add(file);
-                errors.AddRange(ResolveTocModelItems(context, model.Items, parents, file, rootPath, resolveContent, resolveHref, resolveXref));
+                errors.AddRange(ResolveTocModelItems(context, model.Items, parents, file, rootPath, resolveContent, resolveHref, resolveXref, resolveMoniker, rootFileMonikers));
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: we don't need rootFileMonikers anymore

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 8]
```diff
@@ -116,31 +127,59 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
                 var (resolvedTopicHref, resolvedTopicName, document) = ProcessTopicItem(topicUid, topicHref);
 
                 // set resolved href back
                 tocModelItem.Href = resolvedTocHref ?? resolvedTopicHref ?? resolvedTopicItemFromTocHref?.Href;
+                if (resolvedTocHref != null || resolvedTopicHref != null)
+                {
+                    var linkType = UrlUtility.GetLinkType(resolvedTopicHref);
+                    if (linkType == LinkType.External || linkType == LinkType.AbsolutePath)
+                    {
+                        monikers = rootFileMonikers;
```
Comment (OsmondJiang) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: just directlt resolveMonikers(rootPath);

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 6 above)
Comment (928PJY) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: sure

---
In reply to: [293179060](https://github.com/dotnet/docfx/pull/4702#discussion_r293179060) [](ancestors = 293179060)

File: docs/specs/moniker.yml [hunk 9]
```diff
@@ -838,6 +838,37 @@ outputs:
       "metadata": {"monikers": ["netcore-1.1", "netcore-1.2", "netcore-1.3" ]}
     }
 ---
+# Build toc with monikers info - multiple node with same url
+inputs:
+  docfx.yml: |
+    monikerRange:
+      'docs/v1/**': 'netcore-1.0'
+      'docs/v2/**': 'netcore-2.0'
+      'docs/v3/**': 'netcore-1.0 || netcore-2.0'
```
Comment (yufeih) [inline docs/specs/moniker.yml]: For readability
```suggestion
      'docs/v1-v2/**': 'netcore-1.0 || netcore-2.0'
```

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 10]
```diff
@@ -21,10 +21,14 @@ internal static class TableOfContentsParser
 
         public delegate (string content, Document file) ResolveContent(Document relativeTo, SourceInfo<string> href, bool isInclusion);
 
+        public delegate List<string> ResolveMoniker(Document doc);
+
         public static (List<Error> errors, TableOfContentsModel model)
-            Load(Context context, Document file, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref)
+            Load(Context context, Document file, ResolveContent resolveContent, ResolveHref resolveHref, ResolveXref resolveXref, ResolveMoniker resolveMoniker)
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: Consider use `context.DependencyResolver` and `context.MonikerProvider` to save these delegates and parameters. Could be in another PR.
```suggestion
            Load(Context context, Document file)
```

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 5 above)
Comment (yufeih) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: with `TocModelItem { public string[] Monikers { get; set; } = Array.Empty<string>() }`, this could be:

```suggestion
                var childrenMonikers = tocModelItem.Items.SelectMany(child => child.Monikers);
```

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 11]
```diff
@@ -116,31 +121,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
+                        monikers = resolveMoniker(rootPath);
+                    }
+                    else
+                    {
+                        monikers = resolveMoniker(document);
+                    }
+                }
+                else
+                {
+                    tocModelItem.Href = resolvedTopicItemFromTocHref?.Href;
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: remove this line, itis done in line 146

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 12]
```diff
@@ -116,31 +121,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
                 var topicUid = tocModelItem.Uid;
 
+                var monikers = new List<string>();
+
                 var (resolvedTocHref, resolvedTopicItemFromTocHref, subChildren) = ProcessTocHref(tocHref);
                 var (resolvedTopicHref, resolvedTopicName, document) = ProcessTopicItem(topicUid, topicHref);
 
                 // set resolved href back
                 tocModelItem.Href = resolvedTocHref ?? resolvedTopicHref ?? resolvedTopicItemFromTocHref?.Href;
+                if (resolvedTocHref != null || resolvedTopicHref != null)
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: The type resolvedTocHref is `SourceInfo` so:
```suggestion
                if (!string.IsNullOrEmpty(resolvedTocHref) || !string.IsNullOrEmpty(resolvedTopicHref))
```

I need to turn `SourceInfo` into a struct to avoid this problem in the future.

File: src/docfx/build/toc/parser/TableOfContentsParser.cs [hunk 13]
```diff
@@ -116,31 +121,58 @@ private static (List<Error>, TableOfContentsModel) LoadTocModel(JToken tocToken)
...
                 tocModelItem.TocHref = resolvedTocHref;
                 tocModelItem.Homepage = resolvedTopicHref;
                 tocModelItem.Name = tocModelItem.Name ?? resolvedTopicName;
                 tocModelItem.Items = subChildren?.Items ?? tocModelItem.Items;
 
+                // Union with children's monikers
+                var childrenMonikers = tocModelItem.Items?.SelectMany(child => child?.Monikers ?? new List<string>()) ?? new List<string>();
+                monikers = childrenMonikers.Union(monikers).Distinct().ToList();
+                monikers.Sort(context.MonikerProvider.Comparer);
+                tocModelItem.Monikers = monikers;
```
Comment (yufeih) [inline src/docfx/build/toc/parser/TableOfContentsParser.cs]: consider extract a method to contain logics in line 170-174 and line 147-165
```suggestion
                tocModelItem.Monikers = GetMonikers(xxxxx);
```

File: src/docfx/build/toc/parser/TableOfContentsParser.cs (hunk 10 above)
Comment (928PJY) [reply src/docfx/build/toc/parser/TableOfContentsParser.cs]: TODo

---
In reply to: [293207838](https://github.com/dotnet/docfx/pull/4702#discussion_r293207838) [](ancestors = 293207838)

File: docs/specs/moniker.yml (hunk 9 above)
Comment (928PJY) [reply docs/specs/moniker.yml]: OK, fix in next PR

---
In reply to: [293207140](https://github.com/dotnet/docfx/pull/4702#discussion_r293207140) [](ancestors = 293207140)
